internal/testutils/broker.go

```diff
@@ -0,0 +1,307 @@
+}
+
+// BrokerMock is a broker mock that should be used in the tests.
+type BrokerMock struct {
+	isAuthorizedCalls   map[string]isAuthorizedCtx
+	isAuthorizedCallsMu sync.Mutex
+}
+
+// BrokerBusMock is the D-Bus object that will answer calls for the broker mock.
+type BrokerBusMock struct {
+	broker BrokerMock
+}
+
+// StartBusBrokerMock starts the D-Bus service and exports it on the system bus.
+// It returns the configuration file path for the exported broker.
+func StartBusBrokerMock(t *testing.T) string {
+	t.Helper()
+
+	brokerName := strings.ReplaceAll(t.Name(), "/", "_")
+	busObjectPath := fmt.Sprintf(objectPathFmt, brokerName)
+	busInterface := fmt.Sprintf(interfaceFmt, brokerName)
+
+	conn, err := dbus.ConnectSystemBus()
+	require.NoError(t, err, "Setup: could not connect to system bus")
+	t.Cleanup(func() { require.NoError(t, conn.Close(), "Teardown: could not close system bus connection") })
+
+	bus := BrokerBusMock{broker: *NewBrokerMock()}
+	err = conn.Export(&bus, dbus.ObjectPath(busObjectPath), busInterface)
+	require.NoError(t, err, "Setup: could not export mock broker")
+
+	err = conn.Export(introspect.NewIntrospectable(&introspect.Node{
+		Name: busObjectPath,
+		Interfaces: []introspect.Interface{
+			introspect.IntrospectData,
+			{
+				Name:    busInterface,
+				Methods: introspect.Methods(&bus),
+			},
+		},
+	}), dbus.ObjectPath(busObjectPath), introspect.IntrospectData.Name)
+	require.NoError(t, err, "Setup: could not export mock broker introspection")
+
+	reply, err := conn.RequestName(busInterface, dbus.NameFlagDoNotQueue)
+	require.NoError(t, err, "Setup: could not request mock broker name")
+	require.Equal(t, reply, dbus.RequestNameReplyPrimaryOwner, "Setup: mock broker name already taken")
+
+	configPath := writeConfig(t, brokerName)
+
+	t.Cleanup(func() {
+		r, err := conn.ReleaseName(busInterface)
+		require.NoError(t, err, "Teardown: could not release mock broker name")
+		require.Equal(t, r, dbus.ReleaseNameReplyReleased, "Teardown: mock broker name not released")
+	})
+
+	return configPath
+}
+
+func writeConfig(t *testing.T, name string) string {
+	t.Helper()
+
+	cfgPath := filepath.Join(t.TempDir(), "broker-cfg")
+
+	s := fmt.Sprintf(brokerConfigTemplate, name, name, name, name)
+	err := os.WriteFile(cfgPath, []byte(s), 0600)
+	require.NoError(t, err, "Setup: Failed to write broker config file")
+
+	return cfgPath
+}
+
+// NewBrokerMock creates a new fresh broker mock.
+func NewBrokerMock() *BrokerMock {
+	return &BrokerMock{
+		isAuthorizedCalls:   map[string]isAuthorizedCtx{},
+		isAuthorizedCallsMu: sync.Mutex{},
+	}
+}
+
+/* BrokerBusMock implementation */
+
+// NewSession calls the broker mock corresponding method and returns the result or an error requested.
+func (b *BrokerBusMock) NewSession(username, lang string) (sessionID, encryptionKey string, dbusErr *dbus.Error) {
+	if username == "NS_bus_error" {
+		return "", "", dbus.MakeFailedError(errors.New("erroring out as requested"))
```

Comment:
I would add the name of the method in the returned errors. Imagine we have a test chaining multiple calls, maybe the error will be on the first call for NS, where we expected it to pass it and fail on GAM. That way, we can differentiate between each ones.